Support Mod with fcode

moorepants opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->

Fixes #15230 


#### Brief description of what is fixed or changed

Added `Mod: mod` to the list of known Fortran functions.

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* printing
  * Added Mod to the supported Fortran functions.

<!-- END RELEASE NOTES -->

moorepants committed:
Added the Mod() function to the Fortran printer, fixes 15230.

moorepants committed:
Added test for the autowrap command reported in the issue.

moorepants committed:
skip test if f2py isn't installed.

bjodah commented:
My fortran is a bit rusty, so take the following with a grain of salt:

Doesn't `mod` convert arguments to integers? (I see that the test uses floating point, just a bit worried that this might be compiler specific?). I thought `dmod` is for (double precision) floating point numbers. I guess we could use `.is_integer` in `_print_Mod`?

If standard >=95 I think the printer can use `modulo` which is type-generic (?).

moorepants commented:
I'm quite ignorant of Fortran. Just went with obvious, but potentially incorrect, solution. Should probably think more carefully about it. I noticed the c printer had some logic wrt to types.

hsharrison commented:
FWIW the current behavior is correct for an expression with a single mod.

```
expr = x % 2.1
f = autowrap(expr, args=[x])
assert np.isclose(f(2.5), 0.4)
```

Which I think is just writing it as `Mod`. Fortran is case insensitive, right? So think just plain `mod` should work.

I haven't had a chance to test out this branch yet on my full code but I'll do so tomorrow morning.

asmeurer commented:
@certik may also know what the correct usage is here.

moorepants commented:
This https://www.obliquity.com/computer/fortran/intrinsic.html says that mod in Fortran 77 takes either Doulbe, Reals, or Integers.

moorepants commented:
This https://docs.oracle.com/cd/E19422-01/819-3684/3_F77_Intrins.html breaks it down into different functions:
![selection_267](https://user-images.githubusercontent.com/276007/45512333-52dab180-b754-11e8-9101-7c7f9eea00c1.png)

moorepants commented:
More: http://www.public.iastate.edu/~fortran/decfortran-axp/11,67.html:

```
MOD (dividend, divisor)

 A function that divides the first argument by the second and
 returns the remainder.
 +------+---------+----------+------------+-------------+        
 | Args | Generic | Specific |  Argument  | Result Type |
 +------+---------+----------+------------+-------------+
 |  2   |  MOD    |  --      | INTEGER*1  | INTEGER*1   |
 |      |         | IMOD     | INTEGER*2  | INTEGER*2   |
 |      |         | JMOD     | INTEGER*4  | INTEGER*4   |
 |      |see note | KMOD     | INTEGER*8  | INTEGER*8   |
 |      |         | AMOD     | REAL*4     | REAL*4      |
 |      |         | DMOD     | REAL*8     | REAL*8      |
 |      |see note | QMOD     | REAL*16    | REAL*16     |
 +------+---------+----------+------------+-------------+

 NOTE: KMOD is only available on AXP systems.
       QMOD is not available in DEC Fortran.
```

certik commented:
Fortran has two functions, `mod` and `modulo`. Citing from [here](http://computer-programming-forum.com/49-fortran/8841595d70c918ba.htm):

> MOD is defined to be:
>
>   MOD(A,B) = A - int(A/B) * B
>
> while MODULO is defined as
>
>   MODULO(A,B) = A - floor(A/B) * B 

They both correctly work for real or integer.

asmeurer commented:
The difference is that in MOD the sign is the same as A and in MODULO the sign is the same as B. I guess we should default to MODULO since Python's % and SymPy's `Mod` both exhibit that behavior.

bjodah commented:
I think modulo requires fortran 90. But maybe compilers support it for fortran 77 as well?

asmeurer commented:
Would there be any reason to use the specific over the generic? Is MODULO supported by all the Fortran standards we support?

moorepants commented:
I'd try things out but of course my conda gfortran seems to be broken: https://github.com/conda-forge/libgfortran-feedstock/issues/15

moorepants committed:
Switch to modulo and check sign.

certik commented:
Yes, `modulo` is supported in all Fortran compilers (I just tried GFortran, Intel and NAG).

moorepants commented:
`fcode.py` says it prints Fortran 77 code, yet we compile with gfortran in autowrap which doesn't do 77 by default, right?

moorepants committed:
mod -> modulo in tests.

asmeurer commented:
But if someone picks a lower standard in the code generation, it means they only want that standard. So we should give a "not supported in the standard" error if MODULO isn't in the specified standard.

certik commented:
All current Fortran compilers support at least F90, so I suggest we support features that can be compiled with any Fortran compiler in the last 5 years, let's say.

asmeurer commented:
The printer has a `standard` flag that supports several standards: https://github.com/sympy/sympy/blob/4f41286b22a05d5d75f456f1e574d8c115bab5d4/sympy/printing/fcode.py#L132

The default is 77: https://github.com/sympy/sympy/blob/4f41286b22a05d5d75f456f1e574d8c115bab5d4/sympy/printing/fcode.py#L104

We can debate what the default should be (ideally in another issue). I'm not really clear why 77 is the default if most compilers support 90. But regardless, the printer should respect the standard given if MODULO isn't supported and raise an exception (I guess NotImplementedError is used in other places).

moorepants committed:
Merge branch 'master' into issue-15230-fortran-mod

certik commented:
Yes, we can do `NotImplementedError` for 66 and 77. And then in another issue, I suggest to remove 66 and 77.

asmeurer commented:
I don't know if we should remove them, unless they are a large burden to support. At the very least we should try to get some feedback from people before doing so. There could be folks who have to support 77- or 66-only codebases.

moorepants committed:
Fortran Mod() printing raises error if F66 or F77.

- Added _print_Mod() that raises error if invalid Fortran versions.
- Added tests for all standards for Mod() printing.
- Forced autowrap to use the F95 standard to match the statements in the
  docstrings.

moorepants committed:
Added test for integer arguments, removed tempdir.

moorepants commented:
Ok, this has been updated to reflect the conversation. It is working locally for me. Ready for review.

certik reviewed (approved):
That looks good to me. Thanks @moorepants!

moorepants commented on sympy/external/tests/test_autowrap.py:
```diff
@@ -142,6 +142,22 @@ def runtest_issue_10274(language, backend):
+    x, y = symbols('x, y', integer=True)
+    expr = Mod(x, 3) - Mod(y, -2)
+    f = autowrap(expr, args=[x, y], language='F95')
+    exp_res = expr.xreplace({x: 3, y: 2}).evalf()
```
This probably shouldn't evalf() so it stays an integer?

moorepants commented on sympy/printing/fcode.py:
```diff
@@ -304,6 +304,19 @@ def _print_Function(self, expr):
+        if self._settings['standard'] in [66, 77]:
+            msg = ("Python % operator and SymPy's Mod() function are not "
+                   "supported by Fortran 66 or 77 standards.")
+            raise ValueError(msg)
```
Is NotImplementedError preferred?

certik replied:
I think it is.

moorepants commented on sympy/utilities/codegen.py:
```diff
@@ -1204,7 +1204,7 @@ def _call_printer(self, routine):
 
             constants, not_fortran, f_expr = self._printer_method_with_settings(
-                'doprint', dict(human=False, source_format='free'),
+                'doprint', dict(human=False, source_format='free', standard=95),
```
It is not clear to me if the f2py call enforces the use of the F95 compiler. Anybody know?

moorepants committed:
ValueError -> NotImplementedError and remove evalf() to retain integer.

certik commented:
+1 to the last commit.

moorepants committed:
Changed to NotImplementedError in the tests.

moorepants commented:
Ok, tests passing. This can be merged.
